src/middlewared/middlewared/plugins/filesystem.py

```diff
@@ -138,3 +140,57 @@ def file_get_contents(self, path, options=None):
                 f.seek(options['offset'])
             data = binascii.b2a_base64(f.read(options.get('maxlen'))).decode().strip()
         return data
+
+
+class GetAndWatchEventSource(EventSource):
+
+    def run(self):
+        if ':' in self.arg:
+            path, lines = self.arg.rsplit(':', 1)
+            lines = int(lines)
+        else:
+            path = self.arg
+            lines = 3
+        if not os.path.exists(path):
+            # FIXME: Error?
+            return
+
+        bufsize = 8192
+        fsize = os.stat(path).st_size
+        if fsize < bufsize:
+            bufsize = fsize
+        i = 0
+        with open(path) as f:
+            data = []
+            while True:
+                i += 1
+                if bufsize * i > fsize:
+                    break
+                f.seek(fsize - bufsize * i)
+                data.extend(f.readlines())
+                if len(data) >= lines or f.tell() == 0:
+                    break
+
+            self.send_event('ADDED', fields={'data': ''.join(data[-lines:])})
+            f.seek(fsize)
+
+            kqueue = select.kqueue()
+
+            ev = [select.kevent(
```

Comment:
I wonder if we can just abandon this and call `tail` through `subprocess` :)

That way we can also get `-F` for free

Reply:
That would be possible, however I am allergic to subprocess calls =P